src/validators/internationalPhoneNumber.js

```diff
@@ -0,0 +1,56 @@
+import i18n from '@dhis2/d2-i18n'
+import { isEmpty, isString, isNumeric } from './helpers/index.js'
+
+/*
+ * There were some problems with the server side implementation
+ * of how international phone numbers are validated, and the
+ * server side implementation will likely be removed, see:
+ * https://jira.dhis2.org/browse/DHIS2-8040
+ *
+ * So, rather than aligning with the server-side implementation
+ * this validator implements the E.164 numbering plan, see:
+ * https://www.cm.com/blog/how-to-format-international-telephone-numbers/
+ *
+ * SPECS
+ * Here's how the E.164 numbering plan works:
+ * - A telephone number can have a maximum of 15 digits
+ * - The first part of the telephone number is the country code (one to three digits)
+ * - The second part is the national destination code (NDC)
+ * - The last part is the subscriber number (SN)
+ * - The NDC and SN together are collectively called the national (significant) number
+ *
+ * IMPLEMENTATION ADVICE
+ * Two important things to note: First of all, in the international E.164 notation a
+ * leading ‘0’ is removed. The UK mobile phone number ‘07911 123456’ in international
+ * format is ‘+44 7911 123456’, so without the first zero. Secondly in the E.164 notation
+ * all spaces, dashes [‘-‘] and parentheses [ ‘(‘ and ‘)’] are removed, besides the
+ * leading ‘+’ all characters should be numeric.
+ */
```

Comment:
So the current implementation would return undefined for this: `+49 0165 123456`, although it doesn't match the E.164 standard, so this validator does not implement the E.164 correctly.

I think a validator is the wrong solution to this problem, if we want to have #.164 compliant phone number. We actually need an international phone number input component, which has an input or select for the country code and a second input for the number which trims leading zeros and special chars and then checks for a max length of 12.

Reply:
> I think a validator is the wrong solution to this problem, if we want to have #.164 compliant phone number. We actually need an international phone number input component, which has an input or select for the country code and a second input for the number which trims leading zeros and special chars and then checks for a max length of 12.

This part I agree with, in principle. I definitely think that _in the future_ we should get a component like this and once it has been implemented we can probably get rid of this validator. I've have already done some research into this, and there are some excellent external libs that could power this component (i.e. [libphonenumber-js](https://github.com/catamphetamine/libphonenumber-js) or [Google's original libphonenumber](https://github.com/google/libphonenumber)). And the component itself could be a bit like [react-phone-number-input](https://github.com/catamphetamine/react-phone-number-input).

However, while I agree with the above _in principle_, I think that _in practice_ at the moment this validator does add some value. Mainly because we still have a hell of a lot of components to build and I can't imagine that this one would be very high on the priority list. So having a validator in place that validates as little as 80% of cases correctly will be a helpful thing to have.

And this also brings me to the other part of you comment:
> So the current implementation would return undefined for this: +49 0165 123456, although it doesn't match the E.164 standard, so this validator does not implement the E.164 correctly.

I am not claiming that the current algorithm is perfect by a long shot. It only does a very basic check of assessing whether a given value _could be_ a valid phone number:
- It removes characters that might be used a separators (hyphens, dots, spaces)
- It removes leading plus and zero characters
- It check if the cleaned string that is left represents a numerical value
- And if that string does not exceed 15 characters

My goal for this validator was simply to a) reject values that are definitely not valid b) not throw any false positives, i.e. don't reject correct phone numbers.

So, I do think that this validator adds _some_ value. But on the other hand, I can see why perhaps we don't want to add validators that are _roughly right_.

Starting this discussion is a bit of a slippery slope, because to a lesser extend, this would also apply to `email` and `url`. These probably have a much lower likelihood of accepting incorrect values, but they also don't cover 100% of cases.

Looking forward to finding out what you think we should do.

Reply:
BTW: why is `+49 0165 123456` invalid?

Reply:
> BTW: why is +49 0165 123456 invalid?
It is valid.. My bad

<hr />

> Starting this discussion is a bit of a slippery slope, because to a lesser extend, this would also apply to email and url. These probably have a much lower likelihood of accepting incorrect values, but they also don't cover 100% of cases.

@HendrikThePendric and I just quickly talked about this and came to the conclusion: The current way in this PR is fine, most important is: We don't show a warning when a value is correct and we catch most of the cases that are wrong